# Cross Validation

To optimize our model is to perform [Cross Validation](https://en.wikipedia.org/wiki/Cross-validation_(statistics)). It ensure that our model is at its best capacities.

## Importation of dataset, model and all prerequesites

In [ ]:
import joblib
import pandas as pd
from sklearn.preprocessing import OneHotEncoder
from sklearn.model_selection import cross_val_score

In [2]:
loaded_pipeline = joblib.load('../jobs/preprocessing.joblib')
loaded_model = joblib.load('../jobs/logistic_regression_PCA_model.joblib')
data = pd.read_csv('../data/rt_iot2022.csv')

We perform the OneHot Encoder on categorical features.

In [ ]:
X = data.drop('Attack_type', axis=1)
y = data['Attack_type']

categorical_feature = ['proto', 'service']
onehot = OneHotEncoder(sparse_output = False)
encoded_categorical = onehot.fit_transform(data[categorical_feature])

encoded_df = pd.DataFrame(encoded_categorical, columns = onehot.get_feature_names_out(categorical_feature))
X_encoded = X.drop(categorical_feature, axis = 1)
X_encoded = pd.concat([X_encoded, encoded_df], axis = 1)
X_encoded.head()

,id.orig_p,id.resp_p,flow_duration,fwd_pkts_tot,bwd_pkts_tot,fwd_data_pkts_tot,bwd_data_pkts_tot,fwd_pkts_per_sec,bwd_pkts_per_sec,flow_pkts_per_sec,...,service_-,service_dhcp,service_dns,service_http,service_irc,service_mqtt,service_ntp,service_radius,service_ssh,service_ssl
0,38667,1883,32.011598,9,5,3,3,0.281148,0.156193,0.437341,...,0.0,0.0,0.0,0.0,0.0,1.0,0.0,0.0,0.0,0.0
1,51143,1883,31.883584,9,5,3,3,0.282277,0.156821,0.439097,...,0.0,0.0,0.0,0.0,0.0,1.0,0.0,0.0,0.0,0.0
2,44761,1883,32.124053,9,5,3,3,0.280164,0.155647,0.435811,...,0.0,0.0,0.0,0.0,0.0,1.0,0.0,0.0,0.0,0.0
3,60893,1883,31.961063,9,5,3,3,0.281593,0.156440,0.438033,...,0.0,0.0,0.0,0.0,0.0,1.0,0.0,0.0,0.0,0.0
4,51087,1883,31.902362,9,5,3,3,0.282111,0.156728,0.438839,...,0.0,0.0,0.0,0.0,0.0,1.0,0.0,0.0,0.0,0.0


The we compute the remaining of the pre-processing.

In [4]:
X_preprocessed = loaded_pipeline.transform(X_encoded)

## [Cross Validation Score](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.cross_val_score.html)

To be sure that we can get a better result with cross validation method we will evaluate 4 scores with cross validation.

In [ ]:
for score in ['accuracy', 'precision_weighted', 'recall_weighted', 'f1_weighted']:
    cv_scores = cross_val_score(loaded_model, X_preprocessed, y, cv=5, scoring=score)
    print(f"Cross-validation {score} scores: {cv_scores}")
    print(f"Mean {score}: {cv_scores.mean()}")

Cross-validation accuracy scores: [0.94233268 0.95873944 0.98647606 0.99130894 0.98428299]
Mean accuracy: 0.9726280219946817


c:\Users\gaell\Documents\ml_courses\Lib\site-packages\sklearn\metrics\_classification.py:1731: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", result.shape[0])
c:\Users\gaell\Documents\ml_courses\Lib\site-packages\sklearn\metrics\_classification.py:1731: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", result.shape[0])
c:\Users\gaell\Documents\ml_courses\Lib\site-packages\sklearn\metrics\_classification.py:1731: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", result.shape

Cross-validation precision_weighted scores: [0.94142401 0.96517792 0.98686567 0.99114209 0.98061588]
Mean precision_weighted: 0.9730451131645695
Cross-validation recall_weighted scores: [0.94233268 0.95873944 0.98647606 0.99130894 0.98428299]
Mean recall_weighted: 0.9726280219946817
Cross-validation f1_weighted scores: [0.94143925 0.95498598 0.98536092 0.99028304 0.9816519 ]
Mean f1_weighted: 0.9707442177600785


Comparison of baseline model and the mean model after cross validation : 

|           | Base model | Mean of Cross validation models |
|-----------|------------|---------------------------------|
| Accuracy  | 0.977      | 0.973                           |
| Precision | 0.977      | 0.973                           |
| Recall    | 0.977      | 0.973                           |
| F1-score  | 0.976      | 0.971                           |

As we can see after cross validation we lose about 0.4% to 0.5% in all metrics. So using cross validation might not find a better model, our model is for now the best !